samples/matter/lock/src/credentials/credentials_data_types.cpp

```diff
@@ -0,0 +1,179 @@
+/*
+ * Copyright (c) 2024 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include "credentials_data_types.h"
+
+namespace DoorLockData
+{
+
+void pack(void *buff, const void *data, size_t dataSize, size_t &offset)
+{
+	memcpy(reinterpret_cast<uint8_t *>(buff) + offset, data, dataSize);
+	offset += dataSize;
+}
+
+void unpack(void *data, size_t dataSize, const void *buff, size_t &offset)
+{
+	memcpy(data, reinterpret_cast<const uint8_t *>(buff) + offset, dataSize);
+	offset += dataSize;
+}
+
+CHIP_ERROR Credential::FillFromPlugin(EmberAfPluginDoorLockCredentialInfo &credentialInfo)
+{
+	if (credentialInfo.credentialData.size() < kMaxCredentialLength) {
+		mInfo.mFields.mStatus = static_cast<uint8_t>(credentialInfo.status);
+		mInfo.mFields.mCredentialType = static_cast<uint8_t>(credentialInfo.credentialType);
+		mInfo.mFields.mCreationSource = static_cast<uint8_t>(credentialInfo.creationSource);
+		mInfo.mFields.mCreatedBy = static_cast<uint8_t>(credentialInfo.createdBy);
+		mInfo.mFields.mModificationSource = static_cast<uint8_t>(credentialInfo.modificationSource);
+		mInfo.mFields.mLastModifiedBy = static_cast<uint8_t>(credentialInfo.lastModifiedBy);
+		mSecrets.mDataLength = credentialInfo.credentialData.size();
+		memcpy(mSecrets.mData, credentialInfo.credentialData.data(), credentialInfo.credentialData.size());
+		return CHIP_NO_ERROR;
+	}
+	return CHIP_ERROR_BUFFER_TOO_SMALL;
+}
+
+CHIP_ERROR Credential::ConvertToPlugin(EmberAfPluginDoorLockCredentialInfo &credentialInfo) const
+{
+	if (mSecrets.mDataLength < RequiredBufferSize()) {
+		credentialInfo.status = static_cast<DlCredentialStatus>(mInfo.mFields.mStatus);
+		credentialInfo.credentialType = static_cast<CredentialTypeEnum>(mInfo.mFields.mCredentialType);
+		credentialInfo.credentialData = chip::ByteSpan(mSecrets.mData, mSecrets.mDataLength);
+		credentialInfo.creationSource = static_cast<DlAssetSource>(mInfo.mFields.mCreationSource);
+		credentialInfo.createdBy = static_cast<chip::FabricIndex>(mInfo.mFields.mCreatedBy);
+		credentialInfo.modificationSource = static_cast<DlAssetSource>(mInfo.mFields.mModificationSource);
+		credentialInfo.lastModifiedBy = static_cast<chip::FabricIndex>(mInfo.mFields.mLastModifiedBy);
+		return CHIP_NO_ERROR;
+	}
+	return CHIP_ERROR_BUFFER_TOO_SMALL;
+}
+
+size_t Credential::Serialize(void *buff, size_t buffSize)
+{
+	if (!buff || buffSize < RequiredBufferSize() || mSecrets.mDataLength > RequiredBufferSize()) {
+		return 0;
+	}
+	size_t offset = 0;
+
+	pack(buff, mInfo.mRaw, sizeof(mInfo.mRaw), offset);
```

Comment:
fyi, Matter has pretty cool `BufferWriter` and `BufferReader` classes that take care of all the buffer size checks.

Reply:
Yes, I agree. We did this way because we didn't want to use Matter functions there, but I guess it will be nicer to use BufferedWriter and BufferedReader. :)

Reply:
ok, if you're just merging some preexisting code, feel free to ignore such nits. :) This can be cleaned up later.

Reply:
Thanks for the hint, for now let's finalize this PR. I will file a ticket to use Matter utils.